### ISSUE TITLE
No callback received when server moved away for many reason.

### ISSUE THREAD

**Issue by surendrapatel274** [daltoniam/Starscream#817]
I am connected to some web socket server. when server is out of network or somehow not available. there should be some way to get notified that web socket server is not available.

**Comment by ChiellieNL** [comment]
I've noticed the same issue.
When you try to connect to a host that does not exist, the ConnectionState will stay in `.waiting` forever, even after the request's timeoutInterval.

Using: Starscream 4.0.4

To reproduce:
Just set up a connection to `URL(string: "wws://i-dont-exist.xyz")`

**Comment by cs571393** [comment]
I've noticed the same issue.
How to solve this.

**Comment by surendrapatel274** [comment]
What i saw is when socket is disconnected we don't get the away/disconnected callback. But if we do ping server every few seconds as required, then we will get the "Socket Disconnected" actually a timeout callback for the ping we do every few seconds.
Since the timeout of the call is default to around 60 seconds.

**Comment by cs571393** [comment]
@surendrapatel274 
ping server Refers to the socket?.write(ping: Data()) ?

**Comment by surendrapatel274** [comment]
@cs571393  Correct

**Comment by cs571393** [comment]
ws://121.40.165.18:9501
I tried ping the server every 10 seconds, but still callback error.

**Comment by cs571393** [comment]
@surendrapatel274 
ws://121.40.165.18:9501
I tried ping the server every 10 seconds, but still callback error.

**Comment by surendrapatel274** [comment]
> @surendrapatel274
> ws://121.40.165.18:9501
> I tried ping the server every 10 seconds, but still callback error.

Yes, that's correct you will get error callback.. and error callback internally calls forceClose WebSocket connection for that you will get disconnect event. This is a workaround for me, Not the fix.
Well I have tried this with version 4.0.0 and it was working for me.

**Comment by Kerndog73** [comment]
I'm running into the same issue. When the client is disconnected from the server, there is no callback. Is constantly pinging the server the only way to deal with this? Also, if the client disconnects, then tries to send some messages, then reconnects, the messages it tried to send are sent to the server. In some cases this might be desirable but in my case it isn't.

**Comment by KimiChiu** [comment]
There are some minor issues in the 3.x versions. That's why I updated to 4.x. But the lifecycle is really confusing. We were using `websocketDidDisconnect` to deal with all the situations in the old days. But now we have `disconnected` which only be called by server sent close command, `cancelled`, `viabilityChanged` with value `false`, `reconnectSuggested` and `peerClosed`. I'm now using `viabilityChanged` and `peerClosed` to handle connection lost problem.

EDIT

Turns out the mechanic I used above was not working well. It's a lot more complicated. Finally solved all issues after 4 hours.
